libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/ChargingStationsMetadataUpdater.kt

```diff
@@ -0,0 +1,134 @@
+package com.mapbox.navigation.core.reroute
+
+import com.google.gson.JsonObject
+import com.google.gson.JsonPrimitive
+import com.mapbox.api.directions.v5.models.DirectionsResponse
+import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.base.route.RouterOrigin
+import com.mapbox.navigation.core.routeoptions.isEVRoute
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+
+// It's a temporary platform side workaround, the final solution will be on NN side.
+// See https://mapbox.atlassian.net/browse/NN-854
+internal suspend fun restoreChargingStationsMetadata(
+    originalRoute: NavigationRoute,
+    newRoutes: List<NavigationRoute>,
+    workerDispatcher: CoroutineDispatcher = Dispatchers.Default
+): List<NavigationRoute> = withContext(workerDispatcher) {
+    val newPrimaryRoute = newRoutes.first()
+    val rerouteRouteOptions = newPrimaryRoute.routeOptions
+    if (!rerouteRouteOptions.isEVRoute()) {
+        return@withContext newRoutes
+    }
+    val updatedResponseWithoutCorrectTypes = if (newPrimaryRoute.origin == RouterOrigin.Onboard) {
+        updateChargingStationsMetadataBasedOnRequestUrl(rerouteRouteOptions, newPrimaryRoute)
+    } else {
+        newPrimaryRoute.directionsResponse
+    }
+    val updatedResponse = updateChargingStationsTypes(
+        updatedResponseWithoutCorrectTypes,
+        originalRoute.getServerProvidedChargingStationsIds()
+    )
+    // TODO: optimise java memory usage creating a copy of a response NAVAND-1437
+    NavigationRoute.create(updatedResponse, rerouteRouteOptions, newPrimaryRoute.origin)
+}
+
+fun updateChargingStationsTypes(
```

Comment:
I believe we introduce an inconsistency here that may fire after the second reroute.
You change charging station type to server-added here, but you don't adjust reroute options back (you don't remove server-added stations coordiantes from the coordinates list and approachesList, bearingsList, etc.).
So I think we're risking duplicating the coordinates for the second reroute. Also we may use `RouteOptions#coordinatesList` somewhere where we don't expect do see server-added charging stations coordinates.

Reply:
Oh, I guess you added "was_requested_as_user_provided" to fix the issue with the second reroute.
However, I would still vote for restoring options back. Because I'm not sure we (or the end app) won't break if we get `RouteOptions#coordinatesList` with server-added waypoints.

Reply:
> Oh, I guess you added "was_requested_as_user_provided" to fix the issue with the second reroute.

yep.

> However, I would still vote for restoring options back. Because I'm not sure we (or the end app) won't break if we get RouteOptions#coordinatesList with server-added waypoints.

Why do you think we or they should break?

Reply:
Let's check BuildingProcessor, for example.
Also, we don't know how the app uses this method.

Reply:
https://mapbox.atlassian.net/browse/NAVAND-1443